qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -165,9 +187,10 @@ def __init__(
         self._notes = notes or ""
 
         self._jobs = ThreadSafeOrderedDict(job_ids or [])
-        self._job_futures = ThreadSafeList()
-        self._callback_statuses = ThreadSafeOrderedDict()
-        self._callback_future = None
+        self._job_futures = ThreadSafeOrderedDict()
+        self._analysis_callbacks = ThreadSafeOrderedDict()
+        self._analysis_futures = ThreadSafeOrderedDict()
+        self._analysis_executor = futures.ThreadPoolExecutor(max_workers=1)
```

Comment:
One thing is for an analysis executor in a single experiment data we are intentionally running sequential processes, which is why max_workers=1, so here a process pool will not help as far as I can tell. Where this could potentially matter is running analysis in multiple experiments at the same time (ie parallel exp analysis), but im not sure that performance difference would be worth the potential complications. 

In other qiskit modules we have had troubles with process pool on Mac and windows. Also the pickling overhead would need to be measured since this would require the entire expdata to be pickled which may affect performance (and it isn't pickleable by default so some members have to be discarded to pickle, eg see #604). Something that we could explore in future if needed but I don't really think it's needed right now.

Reply:
The general rule of thumb is if you're trying to improve performance on a cpu bound workload (that doesn't require synchronization or shared data) typically using multiprocessing/ProcessPoolExecutor is the path to go. But as @chriseclectic says there are complications with this and it's not exactly a free lunch (you can see a failed attempt on this I tried in terra: https://github.com/Qiskit/qiskit-terra/pull/4781). The platform support in python is not great universally, linux typically works fine but for MacOS we have to disable this by default on Python>=3.8 and for Windows it's disabled by default. There is also the serialization overhead with pickling the objects to consider too and also ensuing everything needed in the child process is pickle-able too . Qiskit as a whole already uses multiprocessing for cpu bound things like `transpile()` or `assemble()` when there is >1 circuit and easily to run in parallel.

However, as I'm reading this PR (I've only skimmed it) we're not really to improve performance of cpu bound analysis but more make the handling of jobs asynchronous and flexible. So have each job run in parallel and waiting on it to return from the backend. A thread pool seems fine for this because we're really just trying to not block the main thread/process while waiting on a job to return and it's more about better scheduling async work. For expensive analysis using a thread pool just means you're not going to actually be running it in parallel, only one thread will have the GIL at a time which doesn't seem any worse performance wise. However, that being said if you do want to use multiprocessing for analysis to try and improve performance nothing is stopping you from doing that in the function you can launch sub processes from a thread if needed (you just have to be careful around tracking handles to the child processes).

Reply:
Fair enough. I was just concerning about the scalability of analysis which is not really tested so far (for example, we may want to use Qiskit Experiments to monitor T1 of eagle chip with 128 qubits with a parallel experiment). Since the process pool sounds super complicated in multi-platform, I agree starting from thread pool is reasonable way to go.